### PR TITLE
core: microfixes

### LINKED ISSUE #1737: Post-preview5 microfixes
Assorted protocol fixes:

neo-project/neo#2293
neo-project/neo#2295
neo-project/neo#2290 (some checks are moved to decoding stage here)
neo-project/neo#2285

### PR DESCRIPTION
**Pull request by AnnaShaleva**
Close #1737

### CODE REVIEW

**Comment by roman-khimov** [comment]
ac527650eb173bb1dbd799d44fea7309fce42ad3 disabled one test, it can be enabled again after 664a2a88eb0e57689c83d83725da769dde809d15.

File: pkg/core/transaction/transaction.go [hunk 1]
```diff
@@ -170,7 +170,11 @@ func (t *Transaction) DecodeBinary(br *io.BinReader) {
 	if br.Err != nil {
 		return
 	}
-	br.ReadArray(&t.Scripts)
+	br.ReadArray(&t.Scripts, len(t.Signers))
+	if len(t.Signers) != len(t.Scripts) {
+		br.Err = fmt.Errorf("number of signers doesn't match witnesses: %d vs %d", len(t.Signers), len(t.Scripts))
```
Comment (roman-khimov) [inline pkg/core/transaction/transaction.go:175]: `ErrTxInvalidWitnessNum` could be moved into `transaction`.

**Comment by AnnaShaleva** [comment]
But IPv6 regexp test still failing (with or without $ at the end).

**Comment by AnnaShaleva** [comment]
> But IPv6 regexp test still failing (with or without $ at the end).

Fixed.

File: pkg/core/native/name_service.go [hunk 2]
```diff
@@ -75,8 +75,8 @@ var (
 	// Lookahead is not supported by Go, but it is simple `(?=.{3,255}$)`,
 	// so we check name length explicitly.
 	nameRegex = regexp.MustCompile("^([a-z0-9]{1,62}\\.)+[a-z][a-z0-9]{0,15}$")
-	ipv4Regex = regexp.MustCompile("^(2(5[0-5]|[0-4]\\d))|1?\\d{1,2}(\\.((2(5[0-5]|[0-4]\\d))|1?\\d{1,2})){3}$")
-	ipv6Regex = regexp.MustCompile("^([a-f0-9A-F]{1,4}:){7}[a-f0-9A-F]{1,4}$")
+	ipv4Regex = regexp.MustCompile("^(?:(?:25[0-5]|2[0-4][0-9]|1[0-9]{2}|[1-9][0-9]|[0-9])\\.){3}(?:25[0-5]|2[0-4][0-9]|1[0-9]{2}|[1-9][0-9]|[0-9])$")
+	ipv6Regex = regexp.MustCompile("(?:^)(([0-9a-f]{1,4}:){7,7}[0-9a-f]{1,4}|([0-9a-f]{1,4}:){1,7}:|([0-9a-f]{1,4}:){1,6}:[0-9a-f]{1,4}|([0-9a-f]{1,4}:){1,5}(:[0-9a-f]{1,4}){1,2}|([0-9a-f]{1,4}:){1,4}(:[0-9a-f]{1,4}){1,3}|([0-9a-f]{1,4}:){1,3}(:[0-9a-f]{1,4}){1,4}|([0-9a-f]{1,4}:){1,2}(:[0-9a-f]{1,4}){1,5}|[0-9a-f]{1,4}:((:[0-9a-f]{1,4}){1,6})|:((:[0-9a-f]{1,4}){1,7}|:))$")
```
Comment (roman-khimov) [inline pkg/core/native/name_service.go:79]: We need more tests for these.

**Review by roman-khimov — APPROVED** [review]

**Pull request merged**